lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,386 @@
+            */
+            mysqlOptions = [
+                "database",
+                "host",
+                "password",
+                "port",
+                "table",
+                "user"
+            ];
+            connectionOptions = {};
+
+            mysqlOptions.forEach((key) => {
+                if (this.config[key]) {
+                    connectionOptions[key] = this.config[key];
+                }
+            });
+
+            this.connection = mysql.createConnection(connectionOptions);
+            this.connection.connect((err) => {
+                if (err) {
+                    throw new Error("There was an error while attempting to connect to the database.");
+                }
+
+                callback(false);
+            });
+        }
+
+
+        /**
+         * Severs the connection from the database.
+         *
+         * @param {Function} callback
+         */
+        disconnect(callback) {
+            this.connection.end(callback);
+        }
+
+
+        /**
+         * Starts polling.
+         *
+         * @param {Function} callback
+         */
+        listen(callback) {
+            this.currentlyPolling = true;
+            poll(this, callback);
+            restore(this);
+        }
+
+
+        /**
+         * Resets the relevant flags and clears the timers, which stops polling.
+         */
+        pausePolling() {
+            timers.clearTimeout(this.poller);
+            timers.clearTimeout(this.restorer);
+            this.poller = null;
+            this.restorer = null;
+            this.currentlyPolling = false;
+        }
+
+
+        /**
+         * Sets the currentlyPolling flag and starts polling.
+         */
+        resumePolling() {
+            this.currentlyPolling = true;
+            poll(this);
+            restore(this);
+        }
+
+
+        /**
+         * Inserts a record to the database. If it's unable to do so because the
+         * record already exists, it will check directly that the entry exists
+         * and if it does, attempt to update the record.
+         *
+         * Inserting a record will repeat as many times as are defined in the
+         * config's createMessageCycleLimit.
+         *
+         * @param {Object} message
+         * @param {string} topics
+         * @param {Function} callback
+         */
+        sendMessage(message, topics, callback) {
+            var cycleCounter, ddqBackendInstance, hash, writeRecord;
+
+            cycleCounter = 0;
+            // eslint-disable-next-line consistent-this
+            ddqBackendInstance = this;
+            hash = crypto.createHash("sha256").update(message).digest("hex");
+
+
+            /**
+             * Tries to create a record. If a record already exists with the
+             * parameters, it will attempt to update that record. Otherwise, it
+             * will call the callback that was provided to sendMessage.
+             */
+            function trySendMessage() {
```

Comment:
... and insert the instance here ...

    function trySendMessage(ddqBackendInstance) {